java-security-test/src/main/java/com/sap/cloud/security/test/JwtGenerator.java

```diff
@@ -158,23 +153,60 @@ public JwtGenerator withClaimValues(String claimName, String... values) {
 	 */
 	public JwtGenerator withClaimsFromFile(String claimsJsonResource) throws IOException {
 		String claimsJson = IOUtils.resourceToString(claimsJsonResource, StandardCharsets.UTF_8);
-		JSONObject claimsAsJsonObject;
-		try {
-			claimsAsJsonObject = new JSONObject(claimsJson);
-		} catch (JSONException e) {
-			throw new JsonParsingException(e.getMessage());
-		}
-		for (String key : claimsAsJsonObject.keySet()) {
-			Object value = claimsAsJsonObject.get(key);
-			jsonPayload.put(key, value);
-		}
+		JSONObject jsonObject = createJsonObject(claimsJson);
+		copyJsonProperties(jsonObject, jsonPayload);
 		return this;
 	}
 
-	private void assertClaimIsSupported(String claimName) {
-		if (unsupportedClaims.contains(claimName)) {
-			throw new UnsupportedOperationException("generic method for claim " + claimName + " is not supported");
+	/**
+	 * This method expects a JSON file that contains data for the token payload and
+	 * header. The file is expected to be in the following format:
+	 * 
+	 * <pre>
+	 * 	"header": {
+	 * 		"alg": "RS256",
+	 * 		"kid": "kid-custom"
+	 *        },
+	 * 	"payload": {
+	 * 		"zid" : "zone-id",
+	 * 		"scope": [
+	 * 			"openid",
+	 * 			"app1.scope"
+	 * 		]
+	 *    }
+	 * </pre>
+	 *
+	 * The payload and header data from the file will be written into the token that
+	 * is being generated. Note that some properties from the file are ignored. This
+	 * is for convenience so that the token can be verified in a test setup rather
+	 * then its original production setup. The following header and payload
+	 * properties are ignored:
+	 * <ul>
+	 * <li>Header: jku, kid</li>
+	 * <li>Payload: exp, aud, iss</li>
+	 * </ul>
+	 *
+	 * If you want to override those fields you need to do so manually after this
+	 * method has been called!
+	 *
+	 * @param tokenJsonResource
+	 *            the resource path to the file containing the json file, e.g.
+	 *            "/token.json"
+	 * @return the builder object.
+	 * @throws JsonParsingException
+	 *             if the file does not contain a valid json object
+	 * @throws IOException
+	 *             when the file cannot be read or does not exist.
+	 */
+	public JwtGenerator fromFile(String tokenJsonResource) throws IOException {
+		String tokenJson = IOUtils.resourceToString(tokenJsonResource, StandardCharsets.UTF_8);
+		JSONObject jsonObject = createJsonObject(tokenJson);
+		copyJsonProperties(filterPayload(jsonObject.optJSONObject("payload")), jsonPayload);
+		copyJsonProperties(filterHeader(jsonObject.optJSONObject("header")), jsonHeader);
+		if (GrantType.CLIENT_CREDENTIALS == getGrantType()) {
```

Comment:
we should discuss that clientId again... related claims: sub, aud, cid, client_id